pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -38,19 +42,138 @@ type Packings struct {
+	// highest pod count.
+	estimator := &packingEstimator{isPodPacked: map[*v1.Pod]bool{}}
+	packings := []*Packings{}
+	for _, pod := range pods {
+		if estimator.isPodPacked[pod] {
+			continue
+		}
+		podsPacked, instances := estimator.calculatePackingForAllTypes(pods)
+		// checked all instance type and found no packing option
+		if len(podsPacked) == 0 || len(instances) == 0 {
+			return nil, fmt.Errorf("no capacity type found for packing %d pods", len(pods))
+		}
+		// keep a track of pods we have packed
+		for _, pod := range podsPacked {
+			estimator.isPodPacked[pod] = true
+		}
+		zap.S().Debugf("For %d pod(s) instance types selected are %v", len(podsPacked), instances)
+		packings = append(packings, &Packings{podsPacked, instances})
+	}
+	return packings, nil
+}
+
+type packingEstimator struct {
+	isPodPacked map[*v1.Pod]bool
+}
+
+func (p *packingEstimator) calculatePackingForAllTypes(pods []*v1.Pod) ([]*v1.Pod, []string) {
+	previousPacking := []*v1.Pod{}
+	instanceTypesSelected := []*nodeCapacity{}
+	// Get all available instance types reverse sorted by their capacity, for every instance
+	// try to fit as many pods as possible. return the list of instances with
+	// highest pods packed and the pods.
+
+	// TODO add filters
+	// TODO reserve (Kubelet+ daemon sets) overhead for instance types
+	// TODO count number of pods created on an instance type
+	for _, instanceOption := range p.getInstanceTypes("") {
+		// check how many pods we can fit on this instance type
+		packings, err := p.calculatePackingForInstance(instanceOption, pods)
+		if err != nil {
+			zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instanceOption.instanceType, err)
+			continue
+		}
+		if len(packings) == 0 {
+			continue
+		}
+		// If the pods packed are the same as before, this instance type can be
+		// considered as a backup option in case we get ICE
+		if podsMatch(previousPacking, packings) {
+			instanceTypesSelected = append(instanceTypesSelected, instanceOption)
+		} else if len(packings) > len(previousPacking) {
+			// If pods packed are more than compared to what we got in last
+			// iteration, consider using this instance type
+			previousPacking = packings
+			instanceTypesSelected = []*nodeCapacity{instanceOption}
+		}
+	}
+	instanceTypeNames := []string{}
+	for _, instance := range instanceTypesSelected {
+		instanceTypeNames = append(instanceTypeNames, instance.instanceType)
+	}
+	return previousPacking, instanceTypeNames
+}
+
+func (p *packingEstimator) calculatePackingForInstance(instance *nodeCapacity, podList []*v1.Pod) ([]*v1.Pod, error) {
+	packing := []*v1.Pod{}
+	// start with the smallest pod based on resources requested
+	for _, pod := range podList {
+		if p.isPodPacked[pod] {
+			continue
+		}
+		cpu := calculateCPURequested(pod)
+		memory := calculateMemoryRequested(pod)
+		if err := instance.reserveCapacity(cpu, memory); err != nil {
+			if errors.Is(err, ErrInsufficientCapacity) {
```

Comment:
I think this shouldn't be modeled as an error - errors are usually for exceptional/unexpected conditions where something is "wrong". Here it seems like we actually expect to encounter this.

Reply:
We need a way to get notified from `reserveCapacity` that it failed to reserver the requested resources. It could be a boolean if it was a success or failure. Do you think a boolean will make more sense here?

Reply:
Yes, IMO error is really only for "something is wrong/unexepcted", but not being able to reserve extra CPU/memory is very much expected.